### PR TITLE
fix edmf bc issue

### PR DESCRIPTION
**Pull request by yairchn**
This PR adds to the prescribed temperature BC the sum of RHS second order fluxes to fluxᵀn, located at the state⁺ level, to match the fluxes at the state⁻ level so that the flux divergence will be zero. This addition was done before only for turbulence model. 

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by yairchn** [comment]
for now I made changes only to the energy BC, in PrescribedTemperature ,  where  a set value (Dirichlet BC) is implemented and the matching of fluxes between state⁻ and state⁺ is needed. As far as I understand that is the only place where these are needed but if people can look and see if they see anything similar in the moisture or momentum bc that would be great

File: src/Atmos/Model/bc_energy.jl [hunk 1]
```diff
@@ -44,21 +44,29 @@ function atmos_energy_boundary_state!(
...
     args,
 )
-    @unpack state⁻, aux⁻, diffusive⁻, t, n⁻ = args
-
-    # TODO: figure out a better way...
-    ν, D_t, _ = turbulence_tensors(atmos, state⁻, diffusive⁻, aux⁻, t)
-    d_h_tot = -D_t .* diffusive⁻.energy.∇h_tot
-    nd_h_tot = dot(n⁻, d_h_tot)
+    @unpack state⁻, aux⁻, diffusive⁻, hyperdiff⁻, t, n⁻ = args
+    tend = Flux{SecondOrder}()
```
Comment (ilopezgp) [inline src/Atmos/Model/bc_energy.jl:56]: `tend_type = Flux{SecondOrder}()` might be better

File: src/Atmos/Model/bc_energy.jl [hunk 2]
```diff
@@ -44,21 +44,29 @@ function atmos_energy_boundary_state!(
...
+    _args = (;
```
Comment (ilopezgp) [inline src/Atmos/Model/bc_energy.jl:57]: `_args⁻` is more specific

File: src/Atmos/Model/bc_energy.jl [hunk 3]
```diff
@@ -44,21 +44,29 @@ function atmos_energy_boundary_state!(
...
+        state = state⁻,
+        aux = aux⁻,
+        t,
+        diffusive = diffusive⁻,
+        hyperdiffusive = hyperdiff⁻,
+    )
+    pargs = merge(_args, (precomputed = precompute(atmos, _args, tend),))
+    total_flux =
+        Σfluxes(Energy(), eq_tends(Energy(), atmos, tend), atmos, pargs)
+    nd_h_tot = dot(n⁻, total_flux)
```
Comment (ilopezgp) [inline src/Atmos/Model/bc_energy.jl:67]: I think `total_flux` is a quantity per unit volume? In that case, can we call this `nd_ρh_tot`?

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 4]
```diff
@@ -898,6 +906,48 @@ function precompute(::EDMF, bl, args, ts, ::Flux{SecondOrder})
...
+            )
+        end,
+    )
+    massflux_h_tot +=
+        (ρa_en * (h_tot_gm - h_tot_en) * (ρu_gm_tup[3] * ρ_inv - env.w))
+
+    ρh_sgs_flux = SVector{3, FT}(
+        0,
+        0,
+        -env.a * K_h * 3 * en_dif.∇h_tot[3] + massflux_h_tot,
```
Comment (ilopezgp) [inline test/Atmos/EDMF/edmf_kernels.jl:947]: Why is there a 3 here? Also, I think ρ is missing.

**Review by charleskawczynski — APPROVED** [review]
This is fine with me, assuming the equations are correct.

File: test/Atmos/EDMF/edmf_kernels.jl (hunk 4 above)
Comment (yairchn) [reply test/Atmos/EDMF/edmf_kernels.jl:947]: you are right about density.
the '3' is here for the vertical component of the gradient as for now we only have vertical components of the SGS flux included

**Comment by yairchn** [comment]
bors try

**Review by ilopezgp — APPROVED** [review]

**Comment by yairchn** [comment]
Bors try

**Comment by yairchn** [comment]
bors try

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 5]
```diff
@@ -898,6 +906,48 @@ function precompute(::EDMF, bl, args, ts, ::Flux{SecondOrder})
...
+    e_tot_up = ntuple(i -> total_energy(e_kin_up[i], _grav * z, ts_up[i]), N_up)
+    h_tot_up = ntuple(i -> total_specific_enthalpy(ts_up[i], e_tot_up[i]), N_up)
+
+    e_tot_en = total_energy(e_kin_en, _grav * z, ts_en)
+    h_tot_en = total_specific_enthalpy(ts_en, e_tot_en)
+    h_tot_gm = total_specific_enthalpy(ts, gm.energy.ρe * ρ_inv)
+
+
+    massflux_h_tot = sum(
+        ntuple(N_up) do i
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_kernels.jl:932]:
```suggestion
        vuntuple(N_up) do i
```

File: test/Atmos/EDMF/edmf_kernels.jl [hunk 6]
```diff
@@ -1034,52 +1085,9 @@ end
...
-                (gm.ρu[3] * ρ_inv - ρaw_up[i] / ρa_up[i]),
-            )
-        end,
-    )
-    massflux_h_tot +=
-        (ρa_en * (h_tot_gm - h_tot_en) * (ρu_gm_tup[3] * ρ_inv - env.w))
-    ρh_sgs_flux = -gm.ρ * env.a * K_h * en_dif.∇h_tot[3] + massflux_h_tot
-    return SVector{3, FT}(0, 0, ρh_sgs_flux)
+    @unpack env, K_h, ρa_up, ρaw_up, ts_up, ts_en, ρh_sgs_flux =
+        args.precomputed.turbconv
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/edmf_kernels.jl:1089]:
```suggestion
    @unpack ρh_sgs_flux = args.precomputed.turbconv
```

**Comment by yairchn** [comment]
bors try

**Comment by yairchn** [comment]
bors try

**Comment by charleskawczynski** [comment]
> Build failed:
> 
> * [buildkite/climatemachine-ci](https://buildkite.com/clima/climatemachine-ci/builds/1703)

cc @yairchn.

**Pull request merged**